File: lectures/lecture-04-why-one-giant-instruction-file-fails/exercises/exercise-01-router-validator/solution/typescript/main.ts
// router-validator exercise, TypeScript solution.
//
// Validates a router-style instruction tree against four structural checks:
// short entry, resolvable routes, hard constraints only in the entry, and
// no rule text duplicated across files. Contract: ../../SPEC.md.

import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";
import { pathToFileURL } from "node:url";

const ENTRY_MAX_LINES = 20;
const RULE_RE = /^- \[([a-z]+)(!?)\] (.+)$/;
const ROUTE_RE = /^- (docs\/[a-z-]+\.md)\b/;

interface Violation {
  readonly file: string;
  readonly line: number;
  readonly detail: string;
}

function readLines(path: string): string[] {
  let lines = readFileSync(path, "utf8").split(/\r?\n/);
  if (lines.length > 0 && lines[lines.length - 1] === "") lines = lines.slice(0, -1);
  return lines;
}

function treeFiles(tree: string): [string, string[]][] {
  const files: [string, string[]][] = [["AGENTS.md", readLines(join(tree, "AGENTS.md"))]];
  const docsDir = join(tree, "docs");
  if (existsSync(docsDir) && statSync(docsDir).isDirectory()) {
    for (const doc of readdirSync(docsDir).filter((f) => f.endsWith(".md")).sort()) {
      files.push([`docs/${doc}`, readLines(join(docsDir, doc))]);
    }
  }
  return files;
}

function checkEntryLength(tree: string): Violation[] {
  const lines = readLines(join(tree, "AGENTS.md"));
  if (lines.length > ENTRY_MAX_LINES) {
    return [{
      file: "AGENTS.md", line: lines.length,
      detail: `entry file is ${lines.length} lines; the router limit is ${ENTRY_MAX_LINES}`,
    }];
  }
  return [];
}

function checkRoutesResolve(tree: string): Violation[] {
  const violations: Violation[] = [];
  readLines(join(tree, "AGENTS.md")).forEach((line, index) => {
    const match = line.match(ROUTE_RE);
    if (match && match[1] && !existsSync(join(tree, match[1]))) {
      violations.push({
        file: "AGENTS.md", line: index + 1,
        detail: `route target does not exist: ${match[1]}`,
      });
    }
  });
  return violations;
}

function checkHardInEntry(tree: string): Violation[] {
  const violations: Violation[] = [];
  for (const [name, lines] of treeFiles(tree)) {
    if (name === "AGENTS.md") continue;
    lines.forEach((line, index) => {
      const match = line.match(RULE_RE);
      if (match && match[2] === "!" && match[3]) {
        violations.push({
          file: name, line: index + 1,
          detail: `hard constraint outside the entry file: ${match[3]}`,
        });
      }
    });
  }
  return violations;
}

function checkNoDuplicates(tree: string): Violation[] {
  const seen = new Map<string, string>();
  const violations: Violation[] = [];
  for (const [name, lines] of treeFiles(tree)) {
    lines.forEach((line, index) => {
      const match = line.match(RULE_RE);
      if (!match || !match[3]) return;
      const text = match[3];
      const earlier = seen.get(text);
      if (earlier) {
        violations.push({
          file: name, line: index + 1,
          detail: `rule text duplicated (also in ${earlier}): ${text}`,
        });
      } else {
        seen.set(text, name);
      }
    });
  }
  return violations;
}

const CHECKS: [string, (tree: string) => Violation[]][] = [
  ["entry-length", checkEntryLength],
  ["routes-resolve", checkRoutesResolve],
  ["hard-in-entry", checkHardInEntry],
  ["no-duplicates", checkNoDuplicates],
];

export function validate(tree: string): Record<string, unknown> {
  const checks = CHECKS.map(([id, run]) => {
    const violations = run(tree);
    return { id, passed: violations.length === 0, violations };
  });
  return { checks, ok: checks.every((check) => check.passed) };
}

function main(argv: readonly string[]): number {
  const tree = argv[2];
  if (!tree || argv.length !== 3) {
    console.error("usage: main.ts <tree-dir>");
    return 2;
  }
  if (!existsSync(join(tree, "AGENTS.md"))) {
    console.error(`error: no AGENTS.md in ${tree}`);
    return 2;
  }
  process.stdout.write(JSON.stringify(validate(tree), null, 2) + "\n");
  return 0;
}

if (import.meta.url === pathToFileURL(process.argv[1] ?? "").href) {
  process.exit(main(process.argv));
}

File: lectures/lecture-04-why-one-giant-instruction-file-fails/exercises/exercise-01-router-validator/solution/python/main.py
"""router-validator exercise, Python solution.

Validates a router-style instruction tree against four structural checks:
short entry, resolvable routes, hard constraints only in the entry, and no
rule text duplicated across files. Contract: ../../SPEC.md.
"""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path

ENTRY_MAX_LINES = 20
RULE_RE = re.compile(r"^- \[([a-z]+)(!?)\] (.+)$")
ROUTE_RE = re.compile(r"^- (docs/[a-z-]+\.md)\b")


def read_lines(path: Path) -> list[str]:
    lines = path.read_text(encoding="utf-8").split("\n")
    if lines and lines[-1] == "":
        lines = lines[:-1]
    return lines


def tree_files(tree: Path) -> list[tuple[str, list[str]]]:
    files = [("AGENTS.md", read_lines(tree / "AGENTS.md"))]
    docs_dir = tree / "docs"
    if docs_dir.is_dir():
        for doc in sorted(docs_dir.glob("*.md")):
            files.append((f"docs/{doc.name}", read_lines(doc)))
    return files


def check_entry_length(tree: Path) -> list[dict]:
    lines = read_lines(tree / "AGENTS.md")
    if len(lines) > ENTRY_MAX_LINES:
        return [{
            "file": "AGENTS.md", "line": len(lines),
            "detail": f"entry file is {len(lines)} lines; the router limit is {ENTRY_MAX_LINES}",
        }]
    return []


def check_routes_resolve(tree: Path) -> list[dict]:
    violations = []
    for number, line in enumerate(read_lines(tree / "AGENTS.md"), 1):
        match = ROUTE_RE.match(line)
        if match and not (tree / match.group(1)).is_file():
            violations.append({
                "file": "AGENTS.md", "line": number,
                "detail": f"route target does not exist: {match.group(1)}",
            })
    return violations


def check_hard_in_entry(tree: Path) -> list[dict]:
    violations = []
    for name, lines in tree_files(tree):
        if name == "AGENTS.md":
            continue
        for number, line in enumerate(lines, 1):
            match = RULE_RE.match(line)
            if match and match.group(2) == "!":
                violations.append({
                    "file": name, "line": number,
                    "detail": f"hard constraint outside the entry file: {match.group(3)}",
                })
    return violations


def check_no_duplicates(tree: Path) -> list[dict]:
    seen: dict[str, str] = {}
    violations = []
    for name, lines in tree_files(tree):
        for number, line in enumerate(lines, 1):
            match = RULE_RE.match(line)
            if not match:
                continue
            text = match.group(3)
            if text in seen:
                violations.append({
                    "file": name, "line": number,
                    "detail": f"rule text duplicated (also in {seen[text]}): {text}",
                })
            else:
                seen[text] = name
    return violations


CHECKS = [
    ("entry-length", check_entry_length),
    ("routes-resolve", check_routes_resolve),
    ("hard-in-entry", check_hard_in_entry),
    ("no-duplicates", check_no_duplicates),
]


def validate(tree: Path) -> dict:
    checks = []
    for check_id, run in CHECKS:
        violations = run(tree)
        checks.append({"id": check_id, "passed": not violations, "violations": violations})
    return {"checks": checks, "ok": all(check["passed"] for check in checks)}


def main(argv: list[str]) -> int:
    if len(argv) != 2:
        print("usage: main.py <tree-dir>", file=sys.stderr)
        return 2
    tree = Path(argv[1])
    if not (tree / "AGENTS.md").is_file():
        print(f"error: no AGENTS.md in {tree}", file=sys.stderr)
        return 2
    print(json.dumps(validate(tree), indent=2))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))
